### DISCUSSION TITLE
Feedback if future Keycloak should use to JDBC_PING discovery in all scenarios

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#29205] [Archive]
## Current status of this discussion

We're trying to understand if JDBC_PING would be a solution, and what people struggle with in this are of Keycloak's configuration. Please provide your feedback below in the comments. We'll then update the the summary below.

## Problem description

The [current cache configuration](https://www.keycloak.org/server/caching) describes multiple JGroups discovery and transport configurations. While for example the Keycloak Operator hides that configuration complexity from users by configuring a reliable setup, there are regularly questions on how to configure discovery in other setups. 

The additional transport stacks for ec2, google and azure require additional libraries that aren't shipped with Keycloak. Users will need to install these in the providers folder to make use of them. 

## Solution idea

Common for all setups is that Keycloak stores its data in the database. So JDBC_PING seems to be a common denominator for all environments. 

## Current gaps

The current Keycloak documentation doesn't describe JDBC ping at all, and the necessary tables are missing in Keycloak's database schema. 

The current configuration doesn't re-use Keycloak's database settings to set up JDBC ping, but requires a separate configuration. Editing Keycloak's Infinispan XML shouldn't be required for such a setup.

Some configurations used by the community:

* https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80

Problems reported by users using JDBC Ping: 

* https://github.com/keycloak/keycloak/issues/10776

Additional difficulties to users are to bind JGroups to the correct host address, and setting the external address for a node: 

* https://github.com/keycloak/keycloak/discussions/10210

Related work: 

* There is a discussion ongoing for a future version of JDBC_PING which might behave better for Keycloak, see https://issues.redhat.com/projects/JGRP/issues/JGRP-2795

**Comment by chrisbra** [comment]
I have been deploying Keycloak in bare deployments in customer environments, so no being able to use Kubernetes features and commonly used jdbc_ping for setting up the infinispan cache. 

Having documentation about how to properly setup jdbc_ping inside Keycloaks project documentation would be very much appreciated. Currently it is often very hairy to set this up correctly and debugging it not easy. Even better would be to have the db ready for this out-of-the box and also show in the UI if Keycloak is running in clustered mode.

Thanks!

**Reply by ahus1** [reply]
> show in the UI if Keycloak is running in clustered mode

Hi @chrisbra - this is a great input, we'll have a look how we can do this.

**Comment by thomasdarimont** [comment]
It is possible to reuse the existing JDBC datasource provided by quarkus in infinispan instead of using a dedicated connection:
try this:
```xml
          <JDBC_PING  stack.combine="REPLACE" stack.position="MPING"
                      datasource_jndi_name="jdbc/datasource"
                      initialize_sql="CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL,cluster_name varchar(200) NOT NULL,updated TIMESTAMP DEFAULT CURRENT_TIMESTAMP,ping_data BYTEA,constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name))"
                      insert_single_sql="INSERT INTO JGROUPSPING (own_addr, cluster_name, ping_data) values (?, ?, ?)"
                      delete_single_sql="DELETE FROM JGROUPSPING WHERE own_addr=? AND cluster_name=?"
                      select_all_pingdata_sql="SELECT ping_data FROM JGROUPSPING WHERE cluster_name=?" />
```

**Reply by ahus1** [reply]
Thank you, this gets us one step closer for configuring the database!

**Reply by thomasdarimont** [reply]
I just checked this and it seems that it no longer works, since the JNDI support is disabled by default in current quarkus versions for security reasons.
See: https://quarkus.io/guides/all-config#quarkus-core_quarkus-naming-enable-jndi

This was to mitigate log4shell and similar attacks. It would be cool if the jdbc_ping would allow to specify something like a "datasource_resolver_class_name".


hypotetical DataSourceResolver interface:
```java
interface DataSourceResolver {

    DataSource getDataSource();
}
```

Simple CDI based implementation that would work right away with Keycloak/Quarkus:
```java
package my.pkg;

public class CdiDataSourceResolver implements DataSourceResolver {
    public DataSource getDataSource() {
        return CDI.current().select(DataSource.class).get();
    }
}
```

or as suggested by Bela:
```java
package my.pkg;

import ...

public class CdiDataSourceResolver implements Supplier<DataSource> {
    public DataSource get() {
        return CDI.current().select(DataSource.class).get();
    }
}
```

hypotetical configuration:
```xml

<JDBC_PING  stack.combine="REPLACE" stack.position="MPING"
    datasource_resolver_class_name="my.pkg.CdiDataSourceResolver"
   ...
/>
```

**Comment by fcoAtWork** [comment]
Hello

I set up multiple environments and always used the JDBC_PING mechanism. I have to admit I could not understand why this mechanism is not part of the documentation, so definitely I support adding it on the official documentation.

**Reply by ahus1** [reply]
Thank you for sharing this. Can you share with us an example configuration you are using? Did you find problems in configuring it, or problems running it in the past?

**Reply by fcoAtWork** [reply]
I took inspiration from this example: https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80
I find it very useful as discovery method and it easily allow to see nodes and their status within the cluster.
I never had issue so far. I installed in production on a RHSSO 7.4 and Maria DB on AWS EC2 instances and on premises KC 22 and Oracle Database. Lately I had an issue in a dev environment on a two-nodes cluster where, if I do not restart the nodes in the correct order, the second node cannot see the other. But this has nothing do I guess with the discovery protocol: the nodes are correctly notified but they communication times out. Still we cannot find the root cause.

**Comment by sschu** [comment]
A few years ago, we used JDBC_PING for the mentioned reasons. At that time, it worked so-so. The biggest problem was that for some reason old entries were not removed correctly. Over time this led to large startup times as each joining node tried to connect the list of nodes in the DB unsuccessfully. However, it could be we did something wrong or the problem is fixed in the meantime. We switched to DNS_PING since we run in Kubernetes and didn't have any problems with that ever since. 
If the switch to JDBC_PING is made, it should definitely be tested what happens if any node leaves/master node leaves/all nodes leave/fail simultaneously.

**Comment by belaban** [comment]
`JDBC_PING` was written around 2010 by Sanne Grinovero, and I guess I maintained it since then, making smaller changes. But - to be honest - it's never seen much love and (IMO) there are way too many options.

Therefore I'm writing `JDBC_PING2`, which will not be based on `FILE_PING`, and will be much simpler. Zombie entries (mentioned by @sschu  above) shouod not be an issue anymore.

Let's use `JDBC_PING` for now, and see what issues we run into with it. Then run with `JDBC_PING2`, and see if those issues are eliminated.

Note btw that a JGroups stack (5.x) can have *multiple discovery protocols*, e.g. `JDBC_PING` and `DNS_PING`, or possibly one `JDBC_PING` connecting to DB1 and another one connecting to DB2.

**Reply by thomasdarimont** [reply]
That sounds very promising. Thanks for the detailed look behind the scenes!

Would you mind taking a look at [this idea](https://github.com/keycloak/keycloak/discussions/29205#discussioncomment-9294239) to make the datasource resolution pluggable?

**Reply by belaban** [reply]
Injecting a `DataSource` via a class which implements something similar to `Supplier<DataSource>` is definitely possible, and should be simple. Noted.

**Reply by pruivo** [reply]
The next Infinispan 15 release will support injecting the data source: https://issues.redhat.com/browse/ISPN-15961

**Comment by xgp** [comment]
Even with the occasional zombie entries as mentioned by @sschu above, we've been using `JDBC_PING` with all of our hosted and customer clusters with very few issues. We've found it to be the easiest to reason and operate. Injecting the existing `DataSource` would make it even easier. Glad to see this is getting some love.

**Reply by belaban** [reply]
I'm not going to deprecate JDBC_PING; it will get some new features, e.g. [1]. At the same time, JDBC_PING2 should be simpler to run, but let the community decide which one will win in the long run.

[1] https://issues.redhat.com/browse/JGRP-2799

**Reply by ahus1** [reply]
@xgp - could you please elaborate on "with very few issues" on what those issues were? Thanks!

**Reply by xgp** [reply]
From memory, we had
1. Issues with database reconnects when the connection was unreliable or we had downtime. We ended up using a local db proxy so that this wouldn't be an issue
2. After testing, we standardized on using `clear_table_on_view_change`, as otherwise crashed members weren't leaving the table.
3. We had to customize `info_writer_max_writes_after_view` and `info_writer_sleep_time` per environment/setup or we'd get inconsistent views.

**Comment by stianst** [comment]
JDBC_PING makes a lot of sense to me for two reasons:

1) Keycloak always has a DB, and this would allow having a single ping mechanisms that should work in all environments making it simpler to setup clustering in all environments
2) There's no need for a timeout waiting for responses from other nodes when there the first node is started. Currently, there is several seconds penalty when starting the first node, for example for development/testing purposes

**Comment by pauliuspetk** [comment]
JDBC_PING or JDBC_PING2 sound great. I'd just add:

- there should be a way to tell Keycloak what hostname/ip and port it should advertise itself as, as to not rely on autodetection mechanisms
- might be a good idea to reuse the existing JDBC connection?

**Comment by belaban** [comment]
@pauliuspetk : if you refer to an existing datasource, then - yes - that can be injected into JDBC_PING{2}, either via JNDI or `datasource_injecter_class`. The latter exists in JDBC_PING and JDBC_PING2, and is a custom class which returns a datasource.
There are mechanisms that allow to bind to the right address, e.g. `match-interface`, `match-address`, `site_local` etc. You can also write a class which picks the interface yourself. By default, JGroups picks the first interface that non-loopback and is `UP`. See [1] for details.
[1] http://www.jgroups.org/manual5/index.html#Transport

**Reply by thomasdarimont** [reply]
Nice! Since when is the `datasource_injecter_class` property available in JGroups? 

@ahus1 I think it would be great if we had a default implementation for a `Function<JDBC_PING,DataSource>` that does `CDI.current().select(DataSource.class).get()` as shown in https://github.com/keycloak/keycloak/discussions/29205#discussioncomment-9294239

**Reply by pruivo** [reply]
You need to inject the data source into Infinispan configuration: https://github.com/pruivo/keycloak/commit/5f75f5f95d4c8aadda290022842e39f9466ebd4e#diff-fe85ff9d09b49c428eb6a66276d3f68fc0d8c329cbb906afd2b59c81ca4e83fcR191

**Reply by thomasdarimont** [reply]
Thanks @pruivo that's also a nice solution, as we wouldn't need a new class to inject the database.

**Reply by pruivo** [reply]
The idea is to use Keycloak to create the table and generate the queries in the right dialect. But, I'm stuck (not a DB expert). 
@ahus1 I can open a PR with the code I have and document how to disable it in case of SQL query issues. A follow-up would be required to make it more robust to other DB vendors. 

ps. `Store IT` with different DBs are green so probably the default SQL queries are fine :shrug:

**Reply by ahus1** [reply]
Providing a link to some branch here "as is" is IMHO all it needs for someone to pick it up. We'll need to wrap up our current tasks and see if this one end up on top in our next planning. If so, we can continue. In the meantime, someone else can give it a go.

**Comment by belaban** [comment]
Hi @thomasdarimont : since 5.3.7 (https://issues.redhat.com/browse/JGRP-2795). If you need this to be available in a different version, it should be pretty painless to backport...

**Reply by thomasdarimont** [reply]
Thanks! Keycloak 25.0.1 is currently using 5.3.6.Final
```
tom@neumann ~/dev/playground/keycloak/keycloak-25.0.1 
$ ll lib/lib/main/ | grep jgroups
-rw-r--r-- 1 tom tom 2,3M 17. Apr 16:12 org.jgroups.jgroups-5.3.6.Final.jar
```
So we would either manually upgrade to 5.3.7 or wait for the next quarkus / infinispan upgrade, what ever happens first for 26.x (or perhaps a 25.x intermediate release)

**Comment by belaban** [comment]
Documentation: http://www.jgroups.org/manual5/index.html#JDBC_PING2

**Comment by belaban** [comment]
@thomasdarimont 5.3.7 is backwards-compatible to 5.3.6, so it should be simple to manually replace the JGroups JAR for testing this feature.

**Reply by thomasdarimont** [reply]
I just gave this a try. Works fine with Keycloak 25.0.1 when I override the jgroups library and use the custom database injector impl: https://gist.github.com/thomasdarimont/63cafa87b417cb5b3e6f9e6e04e64467

**Reply by ahus1** [reply]
Thank you Thomas! @pruivo was giving it a try as well. We would eventually have the table for JGroups managed by Liquibase, and that would require some more wiring to get the right startup order. 

See https://github.com/pruivo/keycloak/tree/t_29399 for the branch. 

We're currently busy with some other bits and pieces... if you want to give it a try with a proper PR, let us know. We might add some more items to the bill :-)

**Reply by thomasdarimont** [reply]
I'll give it a try.  I looked at @pruivo 's branch with the hardcoded JDBC_PING configuration. I think this will work, but if customers want to use a different JGroups configuration they still need to replace the whole infinispan configuration file.

How about moving the jgroups config from the infinispan cache configuration into a separate file?
Perhaps @belaban has some tips how this can be done.

I think the following could work and make it easier for customers to replace the jgroups configuration with their own, by refering to a custom jgroups stack config file via `-Dkeycloak.jgroups.stackfile=jgroups-dns.xml` without having to duplicate the entire infinispan cache config:

ispn-cache.xml:
```xml
<infinispan
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:infinispan:config:15.0 http://www.infinispan.org/schemas/infinispan-config-15.0.xsd"
        xmlns="urn:infinispan:config:15.0">

    <jgroups>
        <stack-file name="keycloak" path="${keycloak.jgroups.stackfile:jgroups-jdbc.xml}"></stack-file>
    </jgroups>

    <cache-container name="keycloak">
        <transport lock-timeout="60000" stack="keycloak"/>
...
```

jgroups-jdbc.xml:
taken from https://github.com/belaban/JGroups/blob/master/conf/jdbc-pg.xml as an example
```xml
<!--
  JDBC_PING2 for Postgresql
-->
<config xmlns="urn:org:jgroups"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:org:jgroups http://www.jgroups.org/schema/jgroups.xsd">
    <TCP
            bind_addr="${jgroups.bind_addr:site_local}"
            bind_port="${jgroups.bind_port:7800}"
            port_range="50"
            recv_buf_size="150000"
            send_buf_size="640000"
            sock_conn_timeout="300ms"
    />

    <JDBC_PING2
            remove_all_data_on_view_change="true"
            register_shutdown_hook="true"
            return_entire_cache="false"
            <!--
            insert_sp="CREATE PROCEDURE deleteAndInsert
                        (addr varchar(200), name varchar(200), cluster varchar(200), ip varchar(200), coord boolean)
                        LANGUAGE SQL
                        BEGIN ATOMIC
                          DELETE from jgroups where address = addr;
                          INSERT INTO jgroups VALUES(addr, name, cluster, ip, coord);
                        END"
            call_insert_sp="call deleteAndInsert(?,?,?,?,?);"
            -->
    />
    <MERGE3 min_interval="10000"
            max_interval="30000"/>
    <FD_SOCK2/>
    <FD_ALL3 timeout="40s" interval="5s" />
    <VERIFY_SUSPECT2 />
    <pbcast.NAKACK2
            use_mcast_xmit="false"
            xmit_interval="500ms"/>
    <UNICAST3
            xmit_interval="500ms"/>
    <pbcast.STABLE
            desired_avg_gossip="5s"
            max_bytes="1000000"/>
    <pbcast.GMS
            print_local_addr="false"
            join_timeout="1s"
            max_join_attempts="1"/>
    <UFC max_credits="2M"
         min_threshold="0.40"/>
    <MFC max_credits="2M"
         min_threshold="0.4"/>
    <FRAG3 frag_size="60000"  />
    <pbcast.STATE_TRANSFER/>
</config>
```

@belaban is it possible to use inheritance of jgroups configurations in jgroups config files? This could help reduce the config quite a bit.
I'm looking for something like the following.
Note the imaginary `extends="tcp"` attribute, which would refer to the `default-jgroups-tcp.xml` config file from the `default-configs` folder on the classpath, to leverage existing infinispan configs.

jgroups-jdbc.xml:
```xml
<config xmlns="urn:org:jgroups"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:org:jgroups http://www.jgroups.org/schema/jgroups-5.3.xsd" 
       extends="tcp">
   <JDBC_PING stack.combine="REPLACE" stack.position="MPING"/>
</config>
```

**Reply by pruivo** [reply]
@thomasdarimont if the user wants to use any other defaults shipped by Infinispan, the user can use `--cache-stack=<stack>` 
https://www.keycloak.org/server/caching#_transport_stacks
Any change outside the default needs a new Infinispan XML configuration file.

**Reply by thomasdarimont** [reply]
Thanks for the reminder. That's used quite often, however I wondered if we can be more flexible here with the groups stack configuration.

**Reply by ahus1** [reply]
IMHO we shouldn't be more flexible with the JGroups stack configuration, as people always struggle here. This thread originally aims to find a solution to resolve 80% problems users have around the disovery, and making the remaining 20% simpler. 

If we don't think JDBC Ping can provide this, we'll need to re-plan. It should get the discovery on the same level of convenience as DNS Ping is proving on the Kubernetes level.

**Comment by sschu** [comment]
I am fine if JDBC_PING is the default in the future but I would not like to remove support for different mechanisms. We have been running with DNS_PING for the last four years without any problem. It works really well in Kubernetes. For us, there is really no reason to change given our experiences so far. We have seen strange issues in Azure SQL (e.g. worker pool exhaustion causing selected queries to fail) so I would rather keep a JGroups discovery mechanism that was 100% reliable for us so far.

**Comment by belaban** [comment]
@thomasdarimont No, JGroups config files don't support inheritance

**Comment by ahus1** [comment]
Keylcoak 26.1 is using the cache-stack `jdbc-ping` by defaut, which under the hood uses JDBC_PING2:
https://github.com/keycloak/keycloak/issues/29399

Thank you for the collaboration! I'm now archiving this issue.
